src/app/modules/dashboard/pages/dashboard/components/cpu-stats/cpu-stats.component.ts

```diff
@@ -0,0 +1,116 @@
+    @ViewChild(BaseChartDirective, {static: false})
+    cpuChart: BaseChartDirective;
+
+    loading$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(true);
+
+    public cpuLineChartData: Array<any> = [0];
+    public cpuLineChartLabels: Array<any> = [''];
+
+    public cpuLineChartOptions: any = {
+        responsive: true,
+        scales: {
+            xAxes: [
+                {
+                    gridLines: {
+                        display: false
+                    }
+                }
+            ],
+            yAxes: [
+                {
+                    id: 'y-0',
+                    position: 'left',
+                    ticks: {
+                        beginAtZero: true,
+                        max: 100,
+                        min: 0
+                    },
+                    gridLines: {
+                        display: false
+                    }
+                }
+            ]
+        }
+    };
+
+    public cpuLineChartColors: Color[] = [
+        {
+            backgroundColor: 'rgba(0,0,0,0)',
+            borderColor: '#2196F3',
+            pointBackgroundColor: 'rgba(148,159,177,1)',
+            pointBorderColor: '#fff',
+            pointHoverBackgroundColor: '#fff',
+            pointHoverBorderColor: 'rgba(148,159,177,0.8)'
+        }
+    ];
+
+    public cpuLineChartLegend = false;
+    public cpuLineChartType = 'line';
+
+    public cpuLineChartDataSet: Array<any> = [
+        {
+            data: this.cpuLineChartData,
+            label: 'CPU Usage (%)',
+            radius: 0
+        }
+    ];
+
+    private subscription: Subscription;
+    private destroy$: Subject<any> = new Subject();
+
+    constructor(private service: DashboardMetricsService,
+                private notify: ToastrService) {
+    }
+
+    getCpuData() {
+        this.service.getCpuUsage()
+            .pipe(takeUntil(this.destroy$))
+            .subscribe(
+                (response: any) => {
+                    let cpuUsage = response.measurements[0].value * 100
+
+                    const _lineChartData = this.cpuLineChartData;
+                    const _lineChartLabels = this.cpuLineChartLabels;
+
+                    if (_lineChartData.length >= 30) {
+                        _lineChartData.shift();
+                        _lineChartLabels.shift();
+                    }
+
+                    _lineChartData.push(cpuUsage);
+
+                    // This is just to be able to render the new data added to the chart
+                    _lineChartLabels.push('');
+
+                    this.cpuLineChartData = _lineChartData;
+                    this.cpuLineChartLabels = _lineChartLabels;
+                    this.cpuChart.chart.update();
+                },
+                (e) => this.notify.error('Failed to complete request')
+        );
+    }
+
+    ngOnInit() {
+        this.subscription = timer(0, 2000).subscribe(val => this.getCpuData());
```

Comment:
FYI, you can avoid storing the subscription by adding a `.pipe(takeUntil(this.destroy$))` which will automatically dispose the subscription when `this.destroy$` completes.

Reply:
Using `timer` here is a good idea! I'm thinking this could be further simplified and improved by using a `mergeMap` and a `map` in a `pipe` before subscribing to the `timer`. You can checkout the [simple-http-polling](https://www.learnrxjs.io/learn-rxjs/recipes/http-polling#example-2-simple-http-polling) example or this [over-simplified one](https://stackblitz.com/edit/rxjs-http-polling-lwhdkd?file=index.ts). Be sure to add the `takeUntil()` at the end in the `pipe()` and also add `takeUntil` into the `mergeMap` (which should be calling `this.service.get{Cpu|Jvm}Usage`). This should prevent memory leaks.

This change should be applied to the other components as well.

Ping me if this is not very clear and I will give you more details.